### PR TITLE
Add testcase for ipify_facts

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Unit test for ipify_facts

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
test/integration/targets/ipify_facts/aliases
test/integration/targets/ipify_facts/tasks/main.yml

##### ANSIBLE VERSION
```
2.4devel
```

### CODE REVIEW

**Comment by abadger** [comment]
Test cases for some platforms are failing with invalid ssl certificate.  Could be differences in the bundled CA certificates on those platforms or differences in the ssl implementation in the python version on those platforms?

**Comment by resmo** [comment]
one workaround would be to add the `validate_certs` param which gets consumed in fetch_url https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/urls.py#L1018 and set it to false for the failing platform.

**Comment by resmo** [comment]
btw, I thought about "mocking" the json HTTP return to reduce the external dependency for testing.

**Comment by resmo** [comment]
> <akasurde> resmo, ping https://github.com/ansible/ansible/pull/26421#issuecomment-313340816, how do I validate_certs:False for failing platform 

there are a few ways. the cleanest would be:

I would create a *target/ipify_facts/var/default.yml* with a `validate_certs: true` and for the platforms failing e.g. *target/ipify_facts/var/MacOSX_10.11.yml* containing `validate_certs: false`.

In your task add the:

~~~yaml
- include_vars: "{{ item }}"
  with_first_found:
   - "{{ ansible_distribution }}_{{ ansible_distribution_version }}.yml"
   - "default.yml"
~~~

The quickest way probably:
~~~yaml
---
# var/main.yml
validate_certs: true
~~~

~~~yaml
- set_fact: 
    validate_certs: false
  when: '"{{ ansible_distribution }}_{{ ansible_distribution_version }}" == "MacOS_10.11" '

- name: get information about current IP using ipify facts
  ipify_facts:
    timeout: 30
    validate_certs: "{{ validate_certs }}"
  register: external_ip

~~~

**Comment by mattclay** [comment]
CI failure in integration tests due to:

```
Unsupported parameters for (ipify_facts) module: validate_certs. Supported parameters include: api_url,timeout
```

File: lib/ansible/modules/net_tools/ipify_facts.py [hunk 1]
```diff
@@ -92,19 +93,20 @@ def run(self):
...
         return result
 
+
 def main():
     global module
     module = AnsibleModule(
-        argument_spec = dict(
+        argument_spec=dict(
             api_url=dict(default='https://api.ipify.org'),
             timeout=dict(type='int', default=10),
```
Comment (resmo) [inline lib/ansible/modules/net_tools/ipify_facts.py]: Great, the only thing to do is to add the param `timeout=dict(type='bool', default=True),` (and the docs for the new param)

**Comment by resmo** [comment]
@Akasurde ping :)

File: lib/ansible/modules/net_tools/ipify_facts.py (hunk 1 above)
Comment (resmo) [reply lib/ansible/modules/net_tools/ipify_facts.py]: I meant `validate_certs=dict(type='bool', default=True),` which is missing,

**Comment by mattclay** [comment]
CI failure in integration tests `test/integration/targets/ipify_facts/tasks/main.yml:27: ipify_facts : get information about current IP using ipify facts`:

```
{
"changed": false, 
"failed": true, 
"msg": "No valid or no response from url https://api.ipify.org within 30 seconds (timeout)"
}
```

**Comment by resmo** [comment]
@mattclay it seem to always fail for "TEST=linux/centos6/2" running in a timeout. I don't see any reason why. Do you have a clue?

**Comment by mattclay** [comment]
@resmo You can reproduce the error locally with:

```
ansible-test integration ipify_facts -v --docker centos6
```

The failure isn't a timeout. The server is immediately returning an error. There is something about the request it doesn't like from the centos6 image.

**Comment by Akasurde** [comment]
@resmo @mattclay The testcase is failing for CentOS 6.9 and Python 2.6. `fetch_url` is failing on Python 2.6 so I would like to skip this testcase for CentOS 6.9. Could you please let me know how can I do that?

**Comment by mattclay** [comment]
@Akasurde Have you figured out why `fetch_url` is failing on python 2.6?

**Comment by Akasurde** [comment]
@mattclay here are some of my findings

`fetch_url` calls `open_url` and internally calls `urllib_request.urlopen`

On Python 2.6 

```
Python 2.6.6 (r266:84292, Aug 18 2016, 15:13:37)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-17)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import urllib2 as urllib_request
>>> url = "https://api.ipify.org?format=json"
>>> urllib_request.urlopen(url).read()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python2.6/urllib2.py", line 126, in urlopen
    return _opener.open(url, data, timeout)
  File "/usr/lib64/python2.6/urllib2.py", line 397, in open
    response = meth(req, response)
  File "/usr/lib64/python2.6/urllib2.py", line 510, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib64/python2.6/urllib2.py", line 435, in error
    return self._call_chain(*args)
  File "/usr/lib64/python2.6/urllib2.py", line 369, in _call_chain
    result = func(*args)
  File "/usr/lib64/python2.6/urllib2.py", line 518, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
urllib2.HTTPError: HTTP Error 400: Bad Request
```

Whereas, it works perfect on Python 2.7 

```
Python 2.7.10 (default, Jul 30 2016, 18:31:42)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import urllib2 as urllib_request
>>> url = "https://api.ipify.org?format=json"
>>> urllib_request.urlopen(url).read()
'{"ip":"*.*.10.110"}'
```

This leaves us with options

1. Stop fixing for Python 2.6 as it is not supported anymore, there by skipping CentOS6 test,
2. Replace `fetch_url` with `requests` in `ipify_facts` module,

Let me know your preference.

**Comment by abadger** [comment]
I remembered something when the Internet was new that I think will solve
this problem.  A url consists of a scheme (https), a hostname (api.ipify.org),
a path, and other optional components like a query string (?format=json).
Your URL is missing the path component.  Many webservers will automatically
insert a missing "/" to make a valid path.  I just tried your test case but
adding the slash between the hostname and the query string and it worked on
python-2.6 so apparently the python-2.6 client library did not add a
missing path automatically but the python-2.7 client library does.  If
you're able to change the url in the test cases to "
https://api.ipify.org/?format=json" then I think this will start working.

**Comment by resmo** [comment]
@abadger Thanks!
@Akasurde lgtm, merging, thanks as well :)

**Comment by Akasurde** [comment]
@abadger Very nice and detailed explanation. Thanks for sharing this information with everyone. Works like a charm !!!

@resmo @mattclay Thanks for review and merge.

**Pull request merged**